src/panels/developer-tools/mqtt/mqtt-subscribe-card.ts

```diff
@@ -0,0 +1,139 @@
+import {
+  LitElement,
+  customElement,
+  TemplateResult,
+  html,
+  property,
+  CSSResult,
+  css,
+} from "lit-element";
+import "@material/mwc-button";
+import "@polymer/paper-input/paper-input";
+import { HomeAssistant } from "../../../types";
+import "../../../components/ha-card";
+
+interface MqttMessage {
+  topic: string;
+  payload: string;
+  qos: number;
+  retain: number;
+}
+
+@customElement("mqtt-subscribe-card")
+class MqttSubscribeCard extends LitElement {
+  @property() public hass?: HomeAssistant;
+
+  @property() private _topic = "";
+
+  @property() private _subscribed?: () => void;
+
+  @property() private _messages: Array<{
+    id: number;
+    message: MqttMessage;
+  }> = [];
+
+  private _messageCount = 0;
+
+  public disconnectedCallback() {
+    super.disconnectedCallback();
+    if (this._subscribed) {
+      this._subscribed();
+      this._subscribed = undefined;
+    }
+  }
+
+  protected render(): TemplateResult {
+    return html`
+      <ha-card header="Listen to a topic">
+        <form>
+          <paper-input
+            .label=${this._subscribed
+              ? "Listening to"
+              : "Topic to subscribe to"}
+            .disabled=${this._subscribed !== undefined}
+            .value=${this._topic}
+            @value-changed=${this._valueChanged}
+          ></paper-input>
+          <mwc-button
+            .disabled=${this._topic === ""}
+            @click=${this._handleSubmit}
+            type="submit"
+          >
+            ${this._subscribed ? "Stop listening" : "Start listening"}
+          </mwc-button>
+        </form>
+        <div class="events">
+          ${this._messages.map(
+            (msg) => html`
+              <div class="event">
+                Message ${msg.id} received:
+                <pre>${JSON.stringify(msg.message, null, 4)}</pre>
+              </div>
+            `
+          )}
+        </div>
+      </ha-card>
+    `;
+  }
+
+  private _valueChanged(ev: CustomEvent): void {
+    this._topic = ev.detail.value;
+  }
+
+  private async _handleSubmit(): Promise<void> {
+    if (this._subscribed) {
+      this._subscribed();
+      this._subscribed = undefined;
+    } else {
+      this._subscribed = await this.hass!.connection.subscribeMessage<
+        MqttMessage
+      >((message) => this._handleMessage(message), {
+        type: "mqtt/subscribe",
+        topic: this._topic,
+      });
+    }
+  }
+
+  private _handleMessage(message: MqttMessage) {
+    const tail =
+      this._messages.length > 30 ? this._messages.slice(0, 29) : this._messages;
+    this._messages = [
+      {
+        message,
```

Comment:
Each payload will come in as a string. Before appending it, we should see if we can parse the payload as JSON and set that on message. Also could we store a time that the message is received?